homeassistant/components/binary_sensor/tod.py

```diff
@@ -0,0 +1,221 @@
+        after = device_config[CONF_AFTER]
+        after_offset = device_config[CONF_AFTER_OFFSET]
+        before = device_config[CONF_BEFORE]
+        before_offset = device_config[CONF_BEFORE_OFFSET]
+        sensors.append(
+            TodSensor(
+                hass, device, after, after_offset, before, before_offset
+            )
+        )
+
+    if not sensors:
+        _LOGGER.error("No sensors added")
+        return
+
+    async_add_entities(sensors)
+    return
+
+
+def is_sun_event(event):
+    """Return true if event is sun event not time."""
+    return event in (SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET)
+
+
+class TodSensor(Entity):
+    """Time of the Day Sensor."""
+
+    def __init__(self, hass, sensor_name, after, after_offset,
+                 before, before_offset):
+        """Init the ToD Sensor..."""
+        self._name = sensor_name
+        self._state = None
+        self.hass = hass
+
+        self._time_before = None
+        self._time_after = None
+        self._after_offset = after_offset
+        self._before_offset = before_offset
+        self._before = before
+        self._after = after
+
+        self._sunrising = None
+        self._sunsetting = None
+        self._next_sunrising = None
+        self._next_sunsetting = None
+
+        self.entity_id = async_generate_entity_id(
+            ENTITY_ID_FORMAT, 'tod_{}'.format(sensor_name), hass=hass)
+        self._calculate_initial_boudary_time()
+
+    def _calculate_initial_boudary_time(self):
+        """Calculate internal absolute time boudaries."""
+        today = self.current_datetime
+
+        if is_sun_event(self._after):
+            after_event_date = \
+                get_astral_event_date(
+                    self.hass, self._after, dt_util.as_utc(today)) or \
+                get_astral_event_next(
+                    self.hass, self._after, dt_util.as_utc(today))
+            if after_event_date:
+                self._time_after = dt_util.as_local(after_event_date)
+            else:
+                _LOGGER.error("After event date for %s unknown", self._after)
+        else:
+            self._time_after = datetime.combine(
+                today.date(), self._after, tzinfo=self.hass.config.time_zone)
+
+        if is_sun_event(self._before):
+            before_event_date = get_astral_event_next(
+                self.hass, self._before, dt_util.as_utc(self._time_after))
+            if before_event_date:
+                self._time_before = dt_util.as_local(before_event_date)
+            else:
+                _LOGGER.error("Before event date for %s unknown", self._before)
+        else:
+            self._time_before = datetime.combine(
+                today.date(), self._before, tzinfo=self.hass.config.time_zone)
+            if self._time_after > self._time_before:
+                self._time_before += timedelta(days=1)
+
+        self._time_after += self._after_offset
+        self._time_before += self._before_offset
```

Comment:
I think there could be issues with daylight savings time here. For example if you add 1 day to March 30th in CET at 2:30, the result will be a time that does not exist! So, depending on how `async_track_point_in_time` works, this will never fire.

Reply:
That's something that need to solve and add the test case.

Reply:
This has been solved by doing all the calculation in UTC.